apps/src/javalab/Playground.js

```diff
@@ -128,38 +128,59 @@ export default class Playground {
   }
 
   addTextItem(itemData) {
-    if (this.isGameOver) {
-      // can't add new items if the game is over
+    if (this.isGameOver || this.itemExists(itemData)) {
+      // can't add new items if the game is over or if the item already exists
       return;
     }
+
+    const textData = {...itemData};
+    delete textData.id;
+    textData.type = PlaygroundItemType.TEXT;
+
+    this.addPlaygroundItem(itemData.id, textData);
   }
 
   removeItem(itemData) {
     if (this.isGameOver) {
       // can't remove items if game is over
       return;
     }
-    if (this.imageItemExists(itemData)) {
+    if (this.itemExists(itemData)) {
       this.removePlaygroundItem(itemData.id);
     }
-    // TODO: handle text deletion
   }
 
   changeItem(itemData) {
     if (this.isGameOver) {
       // can't change items if game is over
       return;
     }
-    if (this.imageItemExists(itemData)) {
-      const newImageData = {...itemData};
+
+    const newItemData = this.getNewItemData(itemData);
+    this.changePlaygroundItem(itemData.id, newItemData);
```

Comment:
should we call this if the item doesn't exist? As playground is written now we can get a change signal for an item that was created on the backend, not added via `addClickableImage` (or other add method), but then had some setter called. Currently we ignore those signals since the item is not on the screen and we will get its updated info if it's ever added.

Reply:
mmm I think we shouldn't -- can you clarify the case you're describing? Do we create items on the backend?

Reply:
Sure, this would be the student code scenario:
```
ImageItem testImage = new ImageItem("image.jpg", 0, 0, 100, 100);
testImage.setX(50);
Playground.board.addImageItem(testImage);
```
Since we send a signal on every `setX`, we get that `setX` signal even though the image hasn't been added to the board. When `addImageItem` is called we will get the updated x value.